parcel-bundler/parcel
# Centralize yarn/npm installs in the main process

Centralize yarn/npm installs in the main process

This:
* Consolidates `PromiseQueue` into a single implementation in utils. The new one written for v2 has a nicer model where promise-returning functions are given to the queue. Expands the type coverage as well.
* Adds `Deferred`s as an option for resolving/rejecting Promises outside the revealing-constructor pattern. This just formalizes what we'd been doing in `PromiseQueue` and other places.
* Refactors `installPackage` to always run package installation in the main thread, queued and deduplicated.
	* In Parcel 1, there were only a couple of callsites and they always made sure to invoke it through `WorkerFarm.callMaster`, but direct calls to `installPackage` slipped through and multiple simultaneous yarn/npm installs began happening and created unpredictable results.
	* We now wrap the (now unified) PromiseQueue to track package requests currently in flight, and if an exact match comes in (e.g. multiple workers request `@babel/core@^7.0.0`), we don't enqueue those. Once an installation has finished, however, it's possible to install the package again.
	* Removes `pipeSpawn`. Since we now have a generic logger, simply forwarding stdout/stderr from the package managers isn't possible. Instead, invoke each with `--json` and parse their output and forward that to the logger.
    	* yarn has a nice streaming, newline-delimited-json api that we parse and adapt to a Node stream and forward that in real-time.
    	* npm outputs a single blob at the end describing the summary of the run. 
	* Note: yarn actually has a `--mutex` options that we should explore using. This should provide an extra lock in the case the user is running the package manager at the same time. For now, enqueueing in the main thread is reasonably safe and works in npm too.

Test Plan: 
* Create a `.babelrc` in the simple example and reference `@babel/preset-env`, but do not install it.
* Run `yarn demo` and verify installation succeeds.
* Remove the babelrc, and run `npm` to create a `package-lock.json`. Restore the babelrc, and repeat step 2.
